File: 1094-car-pooling/1094-car-pooling.cpp
class Solution {
public:
   
    bool carPooling(vector<vector<int>>& trips, int capacity) {
        vector<int>ans(1001);// since trip length is 1<= l <=1000
        for(auto x: trips){ // add passengers at the begining of trip and reduce at each end of trip
         ans[x[1]]+=x[0];
         ans[x[2]]-=x[0];
        }
        for(auto x: ans){ // subtracting capacity from number of passengers at each pt if <0 it exceeds capacity
           capacity-=x;
            if(capacity <0){
                return false;
            }
        }
        return true;
    }
};

File: 1094-car-pooling/1094-car-pooling.py
class Solution:
    def carPooling(self, trips: List[List[int]], capacity: int) -> bool:
        prefix = [0 for _ in range(1002)]
        for num, start, end in trips:
            prefix[start] += num
            prefix[end] -= num
            
        for index in range(1, 1002):
            prefix[index] += prefix[index-1]
            
        for index in range(1002):
            if prefix[index] > capacity:
                return False
            
        return True